### DISCUSSION TITLE
Looking for MAUI Mock Package like Xamarin.Forms.Mocks

### DISCUSSION THREAD

**Discussion by logeshpalani33** [dotnet/maui#18306] [API Review]
### Description

I'm currently working on a project using MAUI and need a mocking framework similar to `Xamarin.Forms.Mocks` for MAUI. However, I'm unsure if there's an equivalent library specifically designed for MAUI. Could anyone recommend an alternative or provide guidance on how to handle mocking in MAUI projects?

### Public API Changes

No API Changes

### Intended Use-Case

For writing test cases on View and View Model level.

**Comment by PureWeen** [comment]
@jonathanpeppers

**Reply by logeshpalani33** [reply]
@jamesmontemagno @jfversluis @samhouts any progress/update on this?

**Reply by jonathanpeppers** [reply]
I don't think you actually need this package for MAUI, see some conversation here: https://github.com/jonathanpeppers/Xamarin.Forms.Mocks/issues/76

**Reply by jsiemensLatitudegeo** [reply]
@jonathanpeppers It still feels necessary. How do you mock DeviceInfo in a test?

Also, is there a way for instantiating UI objects to not blow up in a test, when using "net8.0-windows10.0.19041.0" target framework?

**Reply by jsiemensLatitudegeo** [reply]
eg. 

```
        [TestMethod]
        public void TestMethod1()
        {
            var label = new Label() { Text = "abc" };
        }
```
```
 Message: 
Test method TestProject2.UnitTest1.TestMethod1 threw exception: 
System.TypeInitializationException: The type initializer for 'Microsoft.Maui.Controls.VisualElement' threw an exception. ---> System.TypeInitializationException: The type initializer for 'WinRT.ActivationFactory`1' threw an exception. ---> System.Runtime.InteropServices.COMException: Class not registered (0x80040154 (REGDB_E_CLASSNOTREG))
```

Works fine with 'net8.0' but not 'net8.0-windows10.0.19041.0'.

Are there any internal methods in MAUI that could be used to prevent MAUI from trying to create the underlying platform UI?

**Reply by jonathanpeppers** [reply]
I would try a plain `net8.0` project without the `-windows10.0.19041.0` part.

**Reply by jsiemensLatitudegeo** [reply]
That would work for this particular test, but I have a nuget package dependency that does not support net8.0. Only 'net8.0-windows10.0.19041.0', 'net8.0-ios', and 'net8.0-android'.

**Comment by FreakyAli** [comment]
Honestly i don't see it happening this soon, for the mean time you could checkout https://github.com/dotnet/maui/blob/main/src/Controls/tests/Core.UnitTests/ContentViewUnitTest.cs

**Comment by Moes81** [comment]
I really think this is a really big drawback of Maui. Not being able to mock a `DeviceInfo`, even `IDeviceInfo`, or a `PermissionResult` just feels anachronistic. The need of setting up an emulator in the pipeline just to test simple services, just because an `IDeviceInfo` is injected or another mock returns a `PermissionResult` leaves me speechless. I start to understand why Flutter is popular. The developer experience is just so much smoother.

**Reply by jonathanpeppers** [reply]
Can you share an example of a unit test you'd like to write that is currently impossible? I've asked this question a few times, and no one has given one yet.

**Reply by Moes81** [reply]
I'll provide a little test solution. I just need a little time to create one besides daily business.

**Reply by jsiemensLatitudegeo** [reply]
> Can you share an example of a unit test you'd like to write that is currently impossible? I've asked this question a few times, and no one has given one yet.

@jonathanpeppers 

We instantiate different types of UI dependending on idiom. Here's some example tests from a real project. This is taken from the Xamarin Forms implementation. Trying to migrate to MAUI, but there is no MockForms equivalent.

```
        [TestMethod]
        public async Task CreateContextMenu_WhenCalledOnPhoneAndCanOpenPanel_ReturnsPanelContextMenu()
        {
            MockForms.Init(runtimePlatform: Device.RuntimePlatform, idiom: TargetIdiom.Phone);

            var allOps = AllOperations.Create(new OperationRegistry());
            allOps.PanelOperations.OpenHostPanel.RegisterExecute((comp) => Task.CompletedTask, out _);

            var factory = CreateFactory(allOps:allOps);
            IMapViewProxy mapProxy = await GetMapView();
            Menu menu = new Menu();
            
            var contextMenu = factory.CreateContextMenu(mapProxy, menu);

            Assert.IsTrue(contextMenu is PanelContextMenu);
        }

        [TestMethod]
        public async Task CreateContextMenu_WhenCalledOnDesktop_ReturnsPopupContextMenu()
        {
            MockForms.Init(runtimePlatform: Device.RuntimePlatform, idiom: TargetIdiom.Desktop);

            var factory = CreateFactory();
            IMapViewProxy mapProxy = await GetMapView();
            Menu menu = new Menu();
            
            var contextMenu = factory.CreateContextMenu(mapProxy, menu);

            Assert.IsTrue(contextMenu is PopupContextMenu);
        }
```

**Reply by jonathanpeppers** [reply]
@jsiemensLatitudegeo your example has a lot of code specific to your project, so I'm not seeing the part that isn't possible.

Are you not able to unit test `Menu`? `Map`?

**Reply by jsiemensLatitudegeo** [reply]
@jonathanpeppers The point is that changing the idiom changes the behavior of the code under test. Effectively, the code under test does something like:

```
 if (DeviceInfo.Idiom == Desktop) 
{ 
   // do something
} else {
  // do something else
}
```

The point is that we need to be able to set the idiom in the test in order to test the expected behavior.

**Reply by jonathanpeppers** [reply]
What I would do for now is call this method with System.Reflection:

https://github.com/dotnet/maui/blob/c745e23c755d1d04e1b4325dea16fce8b9a4a98d/src/Essentials/src/DeviceInfo/DeviceInfo.shared.cs#L122-L123

`IDeviceInfo` is public and from essentials, so seems like this should work?

Xamarin.Forms.Mocks didn't have any support for Xamarin.Essentials, and essentials was folded into MAUI.

**Reply by jsiemensLatitudegeo** [reply]
Yes, that will work. Thanks!

**Reply by Moes81** [reply]
What about using `Moq` to mock the state of `IDeviceInfo`? If I do this, i.e. `new Mock<IDeviceInfo>().Setup(m => m.Platform).Returns(DevicePlatform.Android)`, NUnit says "Inconclusive: Test has not run". Any ideas?

**Reply by Moes81** [reply]
Okay, found my problem. This guy on stackoverflow had the exact same issue:
https://stackoverflow.com/questions/72755718/all-tests-are-inconclusive-with-the-same-an-assembly-specified-in-the-applicati

**Comment by Xyncgas** [comment]
ChatGPT is basically the same thing

**Comment by thomasgalliker** [comment]
I came across this discussion when I tried to unit test behaviors (namely, behaviors that attach to ListView). Unit testing requires some extra effor in .NET MAUI apps. To summarize some important points:
- Try to avoid access to static services (e.g. DeviceInfo.Current). Many .NET MAUI sample apps do this - which is probabyl for the sake of simplicity. But when it comes to testing, it's really bad practice! Instead, inject IDeviceInfo into your consumer classes (could be a view model or another service). Register IDeviceInfo in MauiProgram as a singleton dependency.
- Create a unit test project which targets net8.0 (not net8.0-android nor -ios nor -windows). Platform-independent code is tested in a platform-independet unit test project. Have a look at this [sample project](https://github.com/superdevgmbh/PrismMauiApp).
- When you want to test UI related stuff, things become a bit more tricky in .NET MAUI. Example: I want to test if a behavior runs correctly when it's attached to a ListView. In order to achieve this, I had to create and inject a bunch of mock objects to satisfy some MAUI interfaces. I'm sharing this code [here](https://github.com/thomasgalliker/MauiMocks) and in this nuget package: [https://www.nuget.org/packages/Mocks.Maui](https://www.nuget.org/packages/Mocks.Maui)

Furthermore, have a look at how the MAUI guys test their own stuff and how the CommunityToolkit.Maui guys test their stuff: They replace a lot of dependencies with mock objects before they run their test code.

**Reply by Xyncgas** [reply]
Better to find these things out before working on it maybe we need a forum specifically for industrial technologies free of moderation from the people who made the technology

**Comment by ajsuydam** [comment]
I've come across this as I'm trying to write a simple unit test for my view model that uses the `WeakReferenceMessenger `and the `Maui.Storage.FileSystem` functionality. I also have other view models that have Observable properties that break under unit tests with errors that are thrown by `INotifyPropertyChanged `not having its dependencies met.

When using the net8.0 I get an error for the first view model that tells me I need to reference the assembly because it isn't included in the portable binary. I've spent probably three hours at this point researching this and have gotten nowhere. 

For reference, what I have tried so far:
- The process outlined by Gerald Versluis in his youtube video on the subject.
- Referenced #3552 and several other discussions
- Tried targeting` net8.0-windows10.0.19041`
- Adding explicit references to the packages I need

Could someone point me in the right direction? This is my company's first try with MAUI, a lot is riding on this succeeding.
